### DISCUSSION TITLE
Task end event OnPlayerLogin

### DISCUSSION THREAD

**Discussion by L0mchansky** [TownyAdvanced/Towny#6941] [Q&A] [answered]
Good afternoon!

I'm developing my plugin for a public Minecraft server, but I'm not going to distribute it. I need help from more experienced colleagues.

**I got the following problem:**
Based on the analysis of the Towny code, I found out that the OnPlayerLogin task is responsible for registering a player as a resident. (Creating new ones, downloading, updating due to UUID incidents..)
In my plugin, on top of this class, I create another player entity and I need to register my entity when the resident registration is complete. (I don't know how else to make my entity have a reference to the resident).

**Question**
Can you please tell me if there is any event to complete the registration? My analysis turned up nothing. Or will I have to add my own handler to the "Towny" source? (Which I really don't want)

**Comment by LlmDl** [comment, marked as answer]
Event: https://github.com/TownyAdvanced/Towny/blob/master/Towny/src/main/java/com/palmergames/bukkit/towny/event/resident/NewResidentEvent.java

Getting a Resident from a Player: https://github.com/TownyAdvanced/Towny/wiki/TownyAPI#getting-a-resident-from-a-player

**Reply by L0mchansky** [reply]
Thank you. But doesn't NewResidentEvent only fire when absolutely new players register?
If the player has already been previously registered and the server has been restarted, then the resident will load without events.

**Reply by LlmDl** [reply]
I don't know why you would need an event to do what you're asking.

**Reply by L0mchansky** [reply]
I apologize in advance. I don't fully understand how your incredible development works and I want to express my gratitude for such a structured example of a good learning plugin.

As I understand it, the plugin works.

When a player enters the server, if he was not previously registered in the database as a resident, he is registered from scratch. Wrapper class (resident) is created (new player event fired)
If a player enters who has already been registered before, then a wrapper class for this player is created and placed in the plugin instance.

I want to hook on an event when the player's "remembering" process is over. When a resident instance for a given player has been successfully created to wrap it in its own class wrapper.

**Reply by LlmDl** [reply]
> When a player enters the server, if he was not previously registered in the database as a resident, he is registered from scratch. Wrapper class (resident) is created (new player event fired)

We don't make a new Resident because our DB already has a Resident, which was loaded into our Universe when the plugin is started up.

> If a player enters who has already been registered before, then a wrapper class for this player is created and placed in the plugin instance.

Again, the Resident is not re-made brand new if the player re-logs. The Resident already exists and was loaded when the server started.

> I want to hook on an event when the player's "remembering" process is over. When a resident instance for a given player has been successfully created to wrap it in its own class wrapper.

If you were to listen to the PlayerJoinEvent, wait 10-20 ticks in a scheduled task, you can just use TownyAPI.getInstance().getResidents(Player).

Getting a Resident from the Player can be done any time, except during the PlayerJoinEvent which marks a player's first join on the server.

**Reply by L0mchansky** [reply]
Thank you very much! I will experiment :)
